src/drivers/wayland/wayland.c

```diff
@@ -0,0 +1,2555 @@
+    struct wl_buffer* wl_buf = SMM_BUFFER_PROPERTIES(decoration_buffer)->tag[LW_TAG_LOCAL];
+
+    int pos_x, pos_y;
+
+    switch (decoration->type) {
+    case LW_OBJECT_TITLEBAR:
+        pos_x = 0;
+        pos_y = -LW_TITLE_BAR_HEIGHT;
+        break;
+    case LW_OBJECT_BUTTON_CLOSE:
+        pos_x = parent->width - 1 * (LW_BUTTON_MARGIN + LW_BUTTON_SIZE);
+        pos_y = -1 * (LW_BUTTON_MARGIN + LW_BUTTON_SIZE + (LW_BORDER_SIZE / 2));
+        wl_surface_damage(decoration->surface, 0, 0, decoration->width, decoration->height);
+        break;
+#if LV_WAYLAND_XDG_SHELL
+    case LW_OBJECT_BUTTON_MAXIMIZE:
+        pos_x = parent->width - 2 * (LW_BUTTON_MARGIN + LW_BUTTON_SIZE);
+        pos_y = -1 * (LW_BUTTON_MARGIN + LW_BUTTON_SIZE + (LW_BORDER_SIZE / 2));
+        wl_surface_damage(decoration->surface, 0, 0, decoration->width, decoration->height);
+        break;
+    case LW_OBJECT_BUTTON_MINIMIZE:
+        pos_x = parent->width - 3 * (LW_BUTTON_MARGIN + LW_BUTTON_SIZE);
+        pos_y = -1 * (LW_BUTTON_MARGIN + LW_BUTTON_SIZE + (LW_BORDER_SIZE / 2));
+        wl_surface_damage(decoration->surface, 0, 0, decoration->width, decoration->height);
+        break;
+#endif
+    case LW_OBJECT_BORDER_TOP:
+        pos_x = -LW_BORDER_SIZE;
+        pos_y = -(LW_BORDER_SIZE + LW_TITLE_BAR_HEIGHT);
+        break;
+    case LW_OBJECT_BORDER_BOTTOM:
+        pos_x = -LW_BORDER_SIZE;
+        pos_y = parent->height;
+        break;
+    case LW_OBJECT_BORDER_LEFT:
+        pos_x = -LW_BORDER_SIZE;
+        pos_y = -LW_TITLE_BAR_HEIGHT;
+        break;
+    case LW_OBJECT_BORDER_RIGHT:
+        pos_x = parent->width;
+        pos_y = -LW_TITLE_BAR_HEIGHT;
+        break;
+    default:
+        LV_ASSERT_MSG(0, "Invalid object type");
+        return false;
+    }
+
+    if (decoration->subsurface == NULL) {
+        /* Create the subsurface only once */
+        decoration->subsurface = wl_subcompositor_get_subsurface(window->application->subcompositor,
+            decoration->surface,
+            parent->surface);
+        if (!decoration->subsurface) {
+            LV_LOG_ERROR("cannot get subsurface for decoration");
+            goto err_destroy_surface;
+        }
+    }
+
+    wl_subsurface_set_position(decoration->subsurface, pos_x, pos_y);
+    wl_surface_attach(decoration->surface, wl_buf, 0, 0);
+    wl_surface_commit(decoration->surface);
+
+    return true;
+
+err_destroy_surface:
+    wl_surface_destroy(decoration->surface);
+    decoration->surface = NULL;
+
+    return false;
+}
+
+static void detach_wl_decoration(lw_window_t* window,
+    struct lw_surface_object* decoration)
+{
+    if (decoration->subsurface) {
+        wl_subsurface_destroy(decoration->subsurface);
+        decoration->subsurface = NULL;
+    }
+}
+
+static void fill_decoration_color(lv_color32_t* buf, lv_color32_t color, uint32_t px_num)
+{
+    while (px_num > 16) {
+        *buf = color;  buf++; *buf = color; buf++; *buf = color; buf++; *buf = color; buf++;
```

Comment:
Why is it necessary ? Have you profiled the performance benefits of this unwinding ?

Reply:
What if LV_COLOR_DEPTH is set to 16, most wayland compositors/DRM can do 565 natively this only supports 32bit colors

Reply:
LVGL8 has fill_color, but LVGL9 not,  just copied it here...should be better method to do it.  I thought use LVGL to draw it, but I didn't know what to do.   
I thought Just 32bit color supported...but in fact, maybe more color depth need to support.